### PR TITLE
feat(user): show logged in user list

### LINKED ISSUE #106: display list of users logged into application
currently, application creators cannot see the statistics of their application or view reports.

Do the following.
- [ ] open new page by clicking on the "details" button on the application card.
- [ ] Display names of people who have logged into the application.

## Method of submission, one PR for each of the following.
- [ ] Forntend Interface to list details
- [ ] Backend logic to list user names.

### PR DESCRIPTION
**Pull request by tanay13**
fixes #106 {Backend part}

### CODE REVIEW

**Comment by tanay13** [comment]
@YashKumarVerma Please review this

File: src/user/user.service.ts [hunk 1]
```diff
@@ -108,6 +108,10 @@ export class UserService {
   findOneById(userId: string) {
     return this.userModel.findOne({ _id: userId });
   }
+  findUsers(userArray:any)
+  {
+    return this.userModel.find().where('_id').in(userArray)
+  }
```
Comment (YashKumarVerma) [inline src/user/user.service.ts:115]: remove `any` datatype and put exact data type

**Review by YashKumarVerma — APPROVED** [review]

File: src/user/user.service.ts (hunk 1 above)
Comment (tanay13) [reply src/user/user.service.ts:115]: Can you help me out with the data type i tried putting <code>Array<mongoose.Schema.Types.ObjectId></code>but this is also throwing some error?

File: src/dashboard/dashboard.controller.ts [hunk 2]
```diff
@@ -76,6 +76,22 @@ export class DashboardController {
...
+    const user = await this.applicationService.findOneById(id);
+    const records = await this.userService.findUsers(user.participants)
+    var rec = []
+    records.forEach((i)=>{
+      rec.push(i.name)
+      
+    })
+    res.render('dashboard/detail',{rec})
+
+  }
```
Comment (YashKumarVerma) [inline src/dashboard/dashboard.controller.ts:92]: we're using subdocument, use .populate method to get user details

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]

**Comment by tanay13** [comment]
@YashKumarVerma I made the changes, please have look

File: src/application/application.service.ts [hunk 3]
```diff
@@ -46,6 +46,11 @@ export class ApplicationService {
     return this.applicationModel.find();
   }
 
+  findUsers()
+  {
+    return this.applicationModel.find().populate('participants')
+  }
+
```
Comment (YashKumarVerma) [inline src/application/application.service.ts:53]: this is a valid step, but where are we selecting a single application here ?

we need to combine findOneById and .populate to get results in one query

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]
also, add screenshots of application when changes done

File: src/application/application.service.ts (hunk 3 above)
Comment (tanay13) [reply src/application/application.service.ts:53]: After changing the above code to this 

```js
  @Get('/details/:id')
  @UseGuards(JwtAuthGuard)
  async showUser(@Request() req, @Param('id') id: string) {
    const loggedInUser: LoggedInUser = req.user;
    const user = await this.applicationService.findOneById(id);
    const records = await user.populate('participants')
    console.log(records)

  }

```
I am getting this type of result

test1

Do i have to loop through the <code>id</code> array to show the user's detail?

**Comment by tanay13** [comment]
loginuserpart
@YashKumarVerma After the <code>forEach</code> loop when I am printing the array it is showing empty, why is this happening?

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]
the code to display the user list should reside in the dashboard controller and call the service from application module.

I have made the changes required to successfully render the results. Kindly do the necessary changes to get the PR in mergeable state.

**Comment by YashKumarVerma** [comment]
check the latest commit in this branch: https://github.com/UniAuth/uniauth-backend/tree/list-user

**Comment by tanay13** [comment]
loginuserpart
@YashKumarVerma Please review, now its getting rendered properly

File: src/application/application.controller.ts [hunk 4]
```diff
@@ -40,6 +40,16 @@ export class ApplicationController {
...
 
+  //   @Get('/details/:id')
+  //   @UseGuards(JwtAuthGuard)
+  //   async showUser(@Request() req, @Param('id') id: string) {
+  //     const loggedInUser: LoggedInUser = req.user;
+  //     const user = await this.applicationService.findOneById(id);
+  //     const records = await this.applicationService.findUsers()
+  //     console.log(records)
+
+  //   }
```
Comment (YashKumarVerma) [inline src/application/application.controller.ts:51]: can we remove this if this is not required ?

File: src/application/application.service.ts [hunk 5]
```diff
@@ -46,6 +46,11 @@ export class ApplicationService {
     return this.applicationModel.find();
   }
 
+  async findUsers(id: string) {
```
Comment (YashKumarVerma) [inline src/application/application.service.ts:49]: change function name to `findUsersGrantedAccess()`

File: src/dashboard/dashboard.controller.ts [hunk 6]
```diff
@@ -86,4 +86,20 @@ export class DashboardController {
...
+    try {
+      const loggedInUser: LoggedInUser = req.user;
+      const userDetails = await this.applicationService.findUsers(id);
+      res.render('dashboard/details.hbs',{
+        userDetails:userDetails.participants
+      })
+    } catch (e) {
+      res.json({
+        error: true,
+      });
```
Comment (YashKumarVerma) [inline src/dashboard/dashboard.controller.ts:102]: render the error page here

**Review by YashKumarVerma — CHANGES_REQUESTED** [review]
also do not include changes from the yarn.lock file.

Rest LGTM

**Comment by tanay13** [comment]
@YashKumarVerma I made the required changes, but I am not able to remove the yarn.lock from PR

**Comment by YashKumarVerma** [comment]
@tanay13 https://stackoverflow.com/questions/38743912/remove-a-file-from-a-git-pull-request

**Comment by tanay13** [comment]
@YashKumarVerma Thanks for the link, I removed the file please check

**Review by YashKumarVerma — APPROVED** [review]

**Comment by YashKumarVerma** [comment]
thanks !

**Pull request merged**